### PR TITLE
[Model] add implementation of twirls

### PR DESCRIPTION
**Pull request by FFTYYY**
## Description

Add implementation of TWIRLS.
It'll be great if relevant reviewers can be assigned as well: @VoVAllen 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/nn/pytorch/conv/twirlsconv.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+        self.unfolding = UnfoldindAndAttention(self.hidden_d, self.alp, self.lam, self.prop_step, self.attn_aft,
+                                               self.tau, self.T, self.p, self.use_eta, self.init_att, self.attn_dropout, self.precond)
+
+        # if there are really transformations before unfolding, then do init_activate in mlp_aft
+        self.mlp_aft = MLP(self.size_aft_unf, self.hidden_d, self.output_d, self.num_mlp_after,
+                           self.dropout, self.norm,
+                           init_activate=(self.num_mlp_before > 0) and (
+                               self.num_mlp_after > 0)
+                           )
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/twirlsconv.py:160]: My feeling is that `mlp_bef` and `mlp_aft` can be both decoupled:
```python
def __init__(self, ..., mlp_bef, mlp_aft):
    ...
    self.mlp_bef = mlp_bef
    self.mlp_aft = mlp_aft
    ...
```

File: python/dgl/nn/pytorch/conv/twirlsconv.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+def D_power_bias_X(graph, X, power, coeff, bias):
+    """Y = (coeff*D + bias*I)^{power} X"""
+    degs = graph.ndata["deg"]
+    degs = coeff * degs + bias
+    norm = tc.pow(degs, power)
+    Y = X * norm.view(X.size(0), 1)
+    return Y
+
+
+class UnfoldindAndAttention(nn.Module):
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/twirlsconv.py:312]: `Unfolding`

File: python/dgl/nn/pytorch/conv/twirlsconv.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+                                               self.tau, self.T, self.p, self.use_eta, self.init_att, self.attn_dropout, self.precond)
+
+        # if there are really transformations before unfolding, then do init_activate in mlp_aft
+        self.mlp_aft = MLP(self.size_aft_unf, self.hidden_d, self.output_d, self.num_mlp_after,
+                           self.dropout, self.norm,
+                           init_activate=(self.num_mlp_before > 0) and (
+                               self.num_mlp_after > 0)
+                           )
+
+    def forward(self, graph, feat):
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/twirlsconv.py:162]: @VoVAllen Do you want to support sampling (i.e. operating on blocks) in this PR?  It's OK not to.

**Review by BarclayII — CHANGES_REQUESTED** [review]
A couple easy improvements I could think of.

File: python/dgl/nn/pytorch/conv/twirlsconv.py [hunk 4]
```diff
@@ -1,10 +1,8 @@
+# pylint: disable=trailing-whitespace, line-too-long, missing-docstring, invalid-name, useless-super-delegation, comparison-with-itself, no-member
```
Comment (yzh119) [inline python/dgl/nn/pytorch/conv/twirlsconv.py:1]: Why disabling these warnings? You should address these warnings instead of disabling them.

**Review by yzh119 — CHANGES_REQUESTED** [review]

File: python/dgl/nn/pytorch/conv/twirlsconv.py [hunk 5]
```diff
@@ -0,0 +1,660 @@
+"""Torch modules for TWIRLS"""
+# pylint: disable=invalid-name, useless-super-delegation, no-member
```
Comment (BarclayII) [inline python/dgl/nn/pytorch/conv/twirlsconv.py:2]: Ideally we shouldn't have these options.  What was the concrete error message if you remove it?

Comment (FFTYYY) [reply python/dgl/nn/pytorch/conv/twirlsconv.py:2]: 'invalid-name' is for those variable names with only one capital letter (like T), to be consistent with the paper.

'useless-super-delegation' is for some classes that are conceptually subclass of nn.Module but actually have no parameters (some original classes in pytorch also defined this way, like nn.Dropout).

if remove 'no-member', it reports some functions in dgl.nn.function don't exist, but atually they do. BTW it seems all other implementations in dgl.nn.pytorch.conv disabled this option.

**Review by BarclayII — APPROVED** [review]

**Review by yzh119 — APPROVED** [review]
LGTM, @YYY would you mind adding twirl to the docstring? You should edit https://github.com/dmlc/dgl/blob/master/docs/source/api/python/nn.pytorch.rst .
